### PR TITLE
Adding mssql

### PR DESCRIPTION
**Pull request by cliedeman**
Hello, thanks for making this library.

This PR adds support for mssql, a few notes:

- I couldn't do them same nesting  because the error codes do not share the same prefixes.
- The `ForeignKeyViolationError` reports the target table for insert and updates.
- All the errors include the schema `(dbo.table)` but `NotNullViolationError` includes the database too...

Ciaran

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -33,7 +33,7 @@ before_install:
   - sudo service mysql stop || sudo service mysql-5.6 stop || sudo stop mysql-5.6 || sudo stop mysql || echo "failed to stop mysql"
   - sudo docker pull mysql:$MYSQL_VERSION
   - sudo docker run --name mysql -p 3306:3306 -e MYSQL_ROOT_PASSWORD=root -d mysql:$MYSQL_VERSION
-  - sleep 20
```
Comment (koskimas) [inline .travis.yml]: I think this sleep is needed to make the tests run reliably. Or did you so something here that makes the sleep obsolete?

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,26 @@
+version: "3"
```
Comment (koskimas) [inline docker-compose.yml]: Nice! Thanks for writing this

File: .travis.yml (hunk 1 above)
Comment (cliedeman) [reply .travis.yml]: That was an accident sorry, removed

**Comment by koskimas** [comment]
Wow, super nice job! Thank you!

File: tests/ForeignKeyViolationError.js [hunk 3]
```diff
@@ -50,6 +50,11 @@ module.exports = (session) => {
               expect(error.table).to.equal('source');
               expect(error.constraint).to.equal('source_foreign_key_foreign');
             }
+
+            if (session.isMssql()) {
+              expect(error.table).to.equal('dbo.target');
```
Comment (koskimas) [inline tests/ForeignKeyViolationError.js]: This reports different table than on `postgres` and `mysql`. I'm not sure if we should use the same property for it. Also why is the `dbo` prefix there?

File: tests/NotNullViolationError.js [hunk 4]
```diff
@@ -41,6 +41,10 @@ module.exports = (session) => {
             if (session.isPostgres() || session.isSqlite()) {
               expect(error.table).to.equal(table);
             }
+
+            if (session.isMssql()) {
+              expect(error.table).to.equal('master.dbo.' + table);
```
Comment (koskimas) [inline tests/NotNullViolationError.js]: Also here, why is there the `master.dbo` prefix?

File: tests/ForeignKeyViolationError.js (hunk 3 above)
Comment (koskimas) [reply tests/ForeignKeyViolationError.js]: Oh, `dbo` is the schema right? If you want to add that information too, could we add a separate `schema` property?

File: tests/NotNullViolationError.js (hunk 4 above)
Comment (koskimas) [reply tests/NotNullViolationError.js]: So yeah, it's `database.schema`. If you feel that information is needed, I think we should add a separate properties for `database` and `schema`. That way `table` is the same across all databases (when it exists)

File: tests/ForeignKeyViolationError.js (hunk 3 above)
Comment (cliedeman) [reply tests/ForeignKeyViolationError.js]: I left some notes in my first comment. So all the errors have at-least the schema. and one (for whatever reason) includes the database too. 

So I kind of defer to you on how to handle this.

Happy to remove the database from that one specific error and then ensure schema is always split out? Which would mean adding a top level schema field to DBError?

Comment (cliedeman) [reply tests/ForeignKeyViolationError.js]: I didn't add the extra properties initially because it would be null for every single other client

Comment (koskimas) [reply tests/ForeignKeyViolationError.js]: I think it's still better than having the same property that cannot be used the same way.

Comment (koskimas) [reply tests/ForeignKeyViolationError.js]: So to sum up, I think `table` should only hold the table name. If you think `schema` and `database` are useful, add separate properties for them. If not, just ignore the information.

The bigger problem with this particular error is that it points to the `source` table instead of `target` on other databases. Is there a way to dig out the other table from the error message?

Comment (cliedeman) [reply tests/ForeignKeyViolationError.js]: Sounds good to me, will try to save those values.

No unfortunately not, the error only contains the one table.

**Comment by cliedeman** [comment]
Schema and database moved to standalone properties

File: lib/errors/DBError.js [hunk 5]
```diff
@@ -8,6 +8,8 @@ class DBError extends Error {
     this.name = this.constructor.name;
     this.nativeError = args.nativeError;
     this.client = args.client;
+    this.schema = args.schema;
```
Comment (koskimas) [inline lib/errors/DBError.js]: Why did you add these here? No other properties like table, constraint, column etc. are listed here.

Comment (koskimas) [reply lib/errors/DBError.js]: This will cause every DBError to have these fields.

Comment (cliedeman) [reply lib/errors/DBError.js]: That what I understood from this comment: https://github.com/Vincit/db-errors/pull/4#discussion_r268386422.

Should I just drop the fields?

Comment (koskimas) [reply lib/errors/DBError.js]: The fields should be there when available, just as you implemented, but this particular change (the two assignments) adds the fields a with `undefined` values even when they don't even belong to the error type. Every single error would have

```js
schema: undefined
database: undefined
```
properties. There's a difference between an object not having a field and an object having a filed with undefined value, even though accessing them returns `undefined`  in both cases. For example if you `console.log` those objects, you see a different result.

TLDR;
Just remove these assignments here and move them the the error classes that can have those fields. Like here https://github.com/Vincit/db-errors/blob/master/lib/errors/ForeignKeyViolationError.js

Comment (cliedeman) [reply lib/errors/DBError.js]: Got it thanks, Updated the PR and the typescript definitions

**Comment by koskimas** [comment]
Thanks a ton for doing this! I'll release a new version soon.

**Comment by cliedeman** [comment]
> The bigger problem with this particular error is that it points to the source table instead of target on other databases. Is there a way to dig out the other table from the error message?

This is still an issue though. We could add another field to flag which table the error came from?

**Comment by koskimas** [comment]
@cliedeman Let's leave it as is. Those fields are probably just used for debugging, so it shouldn't matter that much. All options here are bad. Let's change that if we get alot of issues at some point.

**Comment by koskimas** [comment]
@cliedeman I just realized that this doesn't run the tests on travis... It's going to take me some weeks to find time to make them work there before I can release.
